src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets

```diff
@@ -110,13 +115,73 @@ Copyright (c) .NET Foundation. All rights reserved.
     ============================================================
                                         CopyFilesToPublishDirectory
 
-    Copy all build outputs, satellites and other necessary files to the publish directory.
+    Copy all build outputs, satellites and other necessary files to the pre-bundle publish directory.
+    Some outputs (ex: PDB files) may be directly copied to the final publish directory, based on settings.
     ============================================================
     -->
   <Target Name="CopyFilesToPublishDirectory"
-          DependsOnTargets="_CopyResolvedFilesToPublishPreserveNewest;
+          DependsOnTargets="_CopyResolvedFilesToPreBundlePublishPreserveNewest;
+                            _CopyResolvedFilesToPreBundlePublishAlways;
+                            _CopyResolvedFilesToPublishPreserveNewest;
                             _CopyResolvedFilesToPublishAlways" />
 
+  <!--
+    ============================================================
+                                       _CopyResolvedFilesToPreBundlePublishPreserveNewest
+
+    Copy _ResolvedFileToPreBundlePublishPreserveNewest items to the pre-bundle publish directory.
+    ============================================================
+    -->
+  <Target Name="_CopyResolvedFilesToPreBundlePublishPreserveNewest"
+          DependsOnTargets="_ComputeResolvedFilesToPublishTypes"
+          Inputs="@(_ResolvedFileToPreBundlePublishPreserveNewest)"
+          Outputs="@(_ResolvedFileToPreBundlePublishPreserveNewest->'$(PreBundlePublishDir)%(RelativePath)')">
+
+    <!--
+        Not using SkipUnchangedFiles="true" because the application may want to change
+        one of these files and not have an incremental build replace it.
+        -->
+    <Copy SourceFiles = "@(_ResolvedFileToPreBundlePublishPreserveNewest)"
+          DestinationFiles="@(_ResolvedFileToPreBundlePublishPreserveNewest->'$(PreBundlePublishDir)%(RelativePath)')"
+          OverwriteReadOnlyFiles="$(OverwriteReadOnlyFiles)"
+          Retries="$(CopyRetryCount)"
+          RetryDelayMilliseconds="$(CopyRetryDelayMilliseconds)"
+          UseHardlinksIfPossible="$(CreateHardLinksForPublishFilesIfPossible)"
+          UseSymboliclinksIfPossible="$(CreateSymbolicLinksForPublishFilesIfPossible)">
+
+      <Output TaskParameter="DestinationFiles" ItemName="FileWrites"/>
```

Comment:
I don't think appending to FileWrites will help here, since incremental clean doesn't write the file back during publish.  Moreover I don't think having an intermediate directory with a copy of all the files is buying you anything.

Reply:
@ericstj: The bundler currently takes in a publish directory and generates the bundled files. This can be changed to accept a series of files. 

However, I'm not clear about one scenario in that case. There was a request to bundle the contents of the publish directory and [all of its subdirectories](https://github.com/dotnet/core-setup/pull/5286#discussion_r264403799).  The bundler currently handles this case: reads files from all sub-directories of the `--source` directory; writes a manifest of files and sub-directory paths into the bundle. 

If there's a `PreBundlePublish` directory, app developers can write anything to it, and have the contents bundled. If we don't have it, I'm not sure how this can be handled. Thanks.

Reply:
In order to support sub-directories, I think the bundler interface needs to support an interface that identifies: 
* The file to be bundled
* The relative path within the bundle (The value of `RelativePath`)